File: 340. Longest Substring with At Most K Distinct Characters.java
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        int[] freq = new int[128];
        int count = 0;
        int ans = 0;
        int left = 0;

        for(int right = 0; right < s.length(); right++){
            freq[s.charAt(right)]++;
            if(freq[s.charAt(right)] == 1) count++;
            if(count > 2 && left < right){
                freq[s.charAt(left)]--;
                if(freq[s.charAt(left++)] == 0) count--;
            }

            ans = Math.max(ans, right - left + 1);
        }

        return ans;
    }
}

File: 340. Longest Substring with At Most K Distinct Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(const std::string& s) {
        std::vector<int> freq(128, 0);
        int count = 0;
        int ans = 0;
        int left = 0;

        for (int right = 0; right < (int)s.size(); right++) {
            freq[s[right]]++;
            if (freq[s[right]] == 1) count++;
            while (count > 2 && left < right) {
                freq[s[left]]--;
                if (freq[s[left]] == 0) count--;
                left++;
            }
            ans = std::max(ans, right - left + 1);
        }

        return ans;
    }
};